flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -350,42 +355,105 @@ private boolean isPartialResourceConfigured(JobGraph jobGraph) {
 
 	private CompletableFuture<Void> persistAndRunJob(JobGraph jobGraph) throws Exception {
 		jobGraphWriter.putJobGraph(jobGraph);
+		return runJob(jobGraph, false);
+	}
+
+	// TODO remove future
+	// same error in recovery -- kill process
+	private CompletableFuture<Void> runJob(JobGraph jobGraph, boolean isRecovery) {
+		Preconditions.checkState(!runningJobs.containsKey(jobGraph.getJobID()));
+		long jobManagerInitializationStarted = System.currentTimeMillis();
+
+		CompletableFuture<JobManagerRunner> jobManagerRunnerFuture = createJobManagerRunner(jobGraph)
+			.thenApplyAsync(
+				FunctionUtils.uncheckedFunction(jobManagerRunner -> {
+					jobManagerRunner.start();
+					return jobManagerRunner;
+				}),
+			getRpcService().getExecutor()); // execute in separate pool to avoid blocking the Dispatcher
+
+		DispatcherJob dispatcherJob;
+		if (isRecovery) {
+			dispatcherJob = DispatcherJob.createForRecovery(jobManagerRunnerFuture);
+		} else {
+			dispatcherJob = DispatcherJob.createForSubmission(jobManagerRunnerFuture, jobGraph, jobManagerInitializationStarted);
+		}
+		runningJobs.put(jobGraph.getJobID(), dispatcherJob);
 
-		final CompletableFuture<Void> runJobFuture = runJob(jobGraph);
-
-		return runJobFuture.whenComplete(BiConsumerWithException.unchecked((Object ignored, Throwable throwable) -> {
-			if (throwable != null) {
-				jobGraphWriter.removeJobGraph(jobGraph.getJobID());
-			}
-		}));
-	}
+		// register handler JM failures & successful job completion
+		final JobID jobId = jobGraph.getJobID();
+		log.debug("reg handler on " + dispatcherJob.getResultFuture());
+		FutureUtils.assertNoException(
+			dispatcherJob.getResultFuture().handleAsync(
+				(ArchivedExecutionGraph archivedExecutionGraph, Throwable throwable) -> {
+					/**
+					 * Cases
+					 *
+					 * A) JobManagerRunner initialization failed
+					 * 	- we have no jobmanager runner instance
+					 * 	- the exception is in the archived exec graph
+					 *
+					 * B) JobManager initialized but failed
+					 *  - we have no archived exec graph, but an exception
+					 *
+					 * C) JobManager returns ArchivedExecutionGraph
+					 *  - store the graph
+					 *
+					 * D) We are not the active JobManagerRunner anymore --> ignore
+					 *
+					 */
```

Comment:
1. Check whether `dispatcherJob == runningJobs.get(jobId)`
2. Handle result
```
if (archivedExecutionGraph != null) {	
							jobReachedGloballyTerminalState(archivedExecutionGraph);	
						} else {	
							final Throwable strippedThrowable = ExceptionUtils.stripCompletionException(throwable);	

							if (strippedThrowable instanceof JobNotFinishedException) {	
								jobNotFinished(jobId);	
							} else {	
								jobMasterFailed(jobId, strippedThrowable);	
							}	
						}
```

No need to access the `getJobManagerRunnerFuture`.